packages/flutter/test/widgets/heroes_test.dart

```diff
@@ -1119,5 +1151,93 @@ void main() {
     expect(tester.getCenter(find.byKey(firstKey)), const Offset(50.0, 50.0));
   });
 
+  testWidgets('Pop interrupts push, reverses flight', (WidgetTester tester) async {
+    await tester.pumpWidget(new MaterialApp(routes: routes));
+    await tester.tap(find.text('twoInset'));
+    await tester.pump(); // begin navigation from / to /twoInset.
+
+    final double epsilon = 0.001;
+    final Duration duration = const Duration(milliseconds: 300);
+
+    await tester.pump();
+    final double x0 = tester.getTopLeft(find.byKey(secondKey)).dx;
+
+    // Flight begins with the secondKey Hero widget lined up with the firstKey widget.
+    expect(x0, 4.0);
 
+    await tester.pump(duration * 0.1);
+    final double x1 = tester.getTopLeft(find.byKey(secondKey)).dx;
+
+    await tester.pump(duration * 0.1);
+    final double x2 = tester.getTopLeft(find.byKey(secondKey)).dx;
+
+    await tester.pump(duration * 0.1);
+    final double x3 = tester.getTopLeft(find.byKey(secondKey)).dx;
+
+    await tester.pump(duration * 0.1);
+    final double x4 = tester.getTopLeft(find.byKey(secondKey)).dx;
+
+    // Pop route /twoInset before the push transition from / to /twoInset has finished.
+    await tester.tap(find.text('pop'));
+
+    // We expect the hero to take the same path as it did flying
+    // from / to /twoInset as it does now, flying from
+    // '/twoInset' back to /.
+
+    await tester.pump();
+    expect(tester.getTopLeft(find.byKey(secondKey)).dx, closeTo(x4, epsilon));
+
+    await tester.pump(duration * 0.1);
+    expect(tester.getTopLeft(find.byKey(secondKey)).dx, closeTo(x3, epsilon));
+
+    await tester.pump(duration * 0.1);
+    expect(tester.getTopLeft(find.byKey(secondKey)).dx, closeTo(x2, epsilon));
+
+    await tester.pump(duration * 0.1);
+    expect(tester.getTopLeft(find.byKey(secondKey)).dx, closeTo(x1, epsilon));
+
+    await tester.pump(duration * 0.1);
+    expect(tester.getTopLeft(find.byKey(secondKey)).dx, closeTo(x0, epsilon));
+
```

Comment:
you should probably add a comment for lines 1190 to 1200 (checking x3...x0) that while this is our current expectation, it's not as important as the test for x4. As in, we definitely don't want x4 to regress, because that would mean the hero lept, but it's not as critical that the path be consistent on the way back, so long as it is smooth and pretty.